### PR TITLE
Fix multiple workflows and add proper engine persistance

### LINKED ISSUE #7127: 🐛 BUG: Fatal uncaught kj::Exception: kj/table.c++:49: failed: inserted row already exists in table while trying to run multiple workflows locally
### Which Cloudflare product(s) does this pertain to?

Workers Runtime, Wrangler, Other

### What version(s) of the tool(s) are you using?

3.83.0

### What version of Node are you using?

_No response_

### What operating system and version are you using?

Linux

### Observed behavior
Wrangler crashes with a kj::Exception when you have more than one workflow setup in wrangler.toml.

### Expected behavior
Both workflows work

### Steps to reproduce
- Setup a worker with two workflows.
- Add both workflows to wrangler.toml
- exception when running in dev

Works:
```
[[workflows]]
name = "photosIngestWorkflow"
binding = "photosIngestWorkflow"
class_name = "photosIngestWorkflow"
```

Does not work, with crash:
```
[[workflows]]
name = "photosIngestWorkflow"
binding = "photosIngestWorkflow"
class_name = "photosIngestWorkflow"

[[workflows]]
name = "photosLooperWorkflow"
binding = "photosLooperWorkflow"
class_name = "photosLooperWorkflow"
```

### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

`
Fatal uncaught kj::Exception: kj/table.c++:49: failed: inserted row already exists in table
stack: /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@5b41f0f /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@322a850 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@322a744 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@2eaf016 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@2eb9126 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@2e898b8 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@2e87295 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@2e75bdf /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@5b3594a /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@5b35d7b /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@5b33cb8 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@5b339c3 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@2e61af3 /lib/x86_64-linux-gnu/libc.so.6@24082 /home/codespace/.npm/_npx/d77349f55c2be1c0/node_modules/@cloudflare/workerd-linux-64/bin/workerd@2e6102d
`

### PR DESCRIPTION
**Pull request by LuisDuarte1**
Fixes #7127.
Closes WOR-322, WOR-395

Previously, engine did store everything in-memory - this is not expected behaviour since R2/KV/DOs all have persistenece, by default. 

Also, multiple workflow definitions were broken since the DO unique key was a constant (therefore it would fail when workerd launched)

---

- Tests
  - [ ] TODO (before merge)
  - [x] Tests included
  - [ ] Tests not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required
  - [ ] Not required because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: there isn't anything to document (except changelog, which will come later) - this is expected behavior

### CODE REVIEW

File: packages/miniflare/src/plugins/workflows/index.ts [hunk 1]
```diff
@@ -62,15 +62,20 @@ export const WORKFLOWS_PLUGIN: Plugin<
 			sharedOptions.workflowsPersist
 		);
 		await fs.mkdir(persistPath, { recursive: true });
-		const storageService: Service = {
-			name: WORKFLOWS_STORAGE_SERVICE_NAME,
+		// each workflow should get its own storage service
```
[the remark this answers is not in the document]
Comment (LuisDuarte1) [reply packages/miniflare/src/plugins/workflows/index.ts:65]: It was personal preference mostly - this way each workflow gets its own folder in `.wrangler/workflow` and makes it way easier to debug the DO state

**Comment by masylum** [comment]
> > Will adding persistence enable workflows to return serializable values like steps do? This would be great to compose workflows.
> 
> I'm sorry, not sure what you mean by that - afaik you shouldn't have any problem composing workflows atm

I meant being able to call a workflow and get the return value. Right now you can only compose by having them triggering each other, like queues. But it would be great to be able to have them act like steps, with a return value. This is how it's implemented at inngest: https://www.inngest.com/docs/reference/functions/step-invoke

**Comment by LuisDuarte1** [comment]
> I meant being able to call a workflow and get the return value. Right now you can only compose by having them triggering each other, like queues. But it would be great to be able to have them act like steps, with a return value. This is how it's implemented at inngest: https://www.inngest.com/docs/reference/functions/step-invoke

That is unrelated to the PR - but you can already do this - you can pass the workflow output as an event payload to the next. What you are describing is syntactic sugar for that, which we might (or might not) consider in the future :smile:  thanks ^^

File: fixtures/workflow-multiple/tests/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+describe("Workflows", () => {
+	let ip: string,
+		port: number,
+		stop: (() => Promise<unknown>) | undefined,
+		getOutput: () => string;
+
+	beforeAll(async () => {
+		// delete previous run contents because of persistence
+		await rm(resolve(__dirname, "..") + "/.wrangler", {
```
[the remark this answers is not in the document]
Comment (LuisDuarte1) [reply fixtures/workflow-multiple/tests/index.test.ts:15]: Done :)

File: .changeset/cyan-geese-smell.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+"@cloudflare/workflows-shared": minor
+"miniflare": patch
```
Comment (emily-shen) [inline .changeset/cyan-geese-smell.md:3]: nit: think this should be a minor for miniflare

Comment (LuisDuarte1) [reply .changeset/cyan-geese-smell.md:3]: yeah that's fair, i'll change it

File: packages/miniflare/test/plugins/workflows/index.spec.ts
Comment (emily-shen) [inline packages/miniflare/test/plugins/workflows/index.spec.ts:1]: <3 thanks

**Review by emily-shen — APPROVED** [review]
looks good from the wrangler side pending a tiny fixup

**Pull request merged**